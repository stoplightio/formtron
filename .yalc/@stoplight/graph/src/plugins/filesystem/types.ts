import { Stats } from 'fs';

import { INodeInstance, INodeOptions } from '../../types';

export enum FilesystemTypes {
  DIRECTORY = 'directory',
  FILE = 'file',
}

export interface IFilesystemPluginOpts {
  recursive?: boolean;
}

export interface IDirectoryInput extends INodeOptions<IDirectoryContent> {
  type: FilesystemTypes.DIRECTORY;
  path: string;
  stats?: Stats;
}

export interface IDirectory extends INodeInstance<IDirectoryContent> {
  type: FilesystemTypes.DIRECTORY;
  path: string;
  name: string;
  dirname: string;
}

export type IDirectoryContent = Array<{
  name: string;
  path: string;
  stats: Stats;
}>;

export type IFileContent = string;

export interface IFileInput extends INodeOptions<IFileContent> {
  type: FilesystemTypes.FILE;
  path: string;
  stats?: Stats;
}

export interface IFile extends INodeInstance<IFileContent> {
  type: FilesystemTypes.FILE;
  path: string;
  name: string;
  fileName: string;
  dirname: string;
  contentType: string;
  stats?: Stats;
}
