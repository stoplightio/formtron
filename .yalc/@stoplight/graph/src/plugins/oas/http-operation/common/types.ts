import {
  IApiKeySecurityScheme,
  IBasicSecurityScheme,
  IBearerSecurityScheme,
  IOauth2SecurityScheme,
  IOpenIdConnectSecurityScheme,
} from '@stoplight/types';
import { OpenAPIObject, OperationObject, PathObject } from 'openapi3-ts';
import { Operation, Path, Spec } from 'swagger-schema-official';

export declare type HttpSecurityScheme =
  | IApiKeySecurityScheme
  | IBearerSecurityScheme
  | IBasicSecurityScheme
  | IOauth2SecurityScheme
  | IOpenIdConnectSecurityScheme;

export interface ITranslateOperationOptions<O, P, S> {
  operation: O;
  method: string;
  pathObject: P;
  path: string;
  spec: S;
}

export type Oas2TranslateOperationOptions = ITranslateOperationOptions<Operation, Path, Spec>;
export type Oas3TranslateOperationOptions = ITranslateOperationOptions<OperationObject, PathObject, OpenAPIObject>;
