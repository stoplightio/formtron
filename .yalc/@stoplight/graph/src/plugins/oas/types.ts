export enum Oas2RootNodeTypes {
  OAS2 = 'oas2',
  SECURITY_SCHEMES = 'oas2_security_schemes',
  PATHS = 'oas2_paths',
  TAGS = 'oas2_tags',
  SHARED = 'oas2_shared',
}

export enum Oas2ChildNodeTypes {
  SECURITY_SCHEME = 'oas2_security_scheme',
  PATH = 'oas2_path',
  TAG = 'oas2_tag',
}

type OasBaseParamType = 'header' | 'query' | 'path';

export type Oas2ParamType = OasBaseParamType | 'body';
export type Oas3ParamType = OasBaseParamType | 'cookie';

export enum Oas3RootNodeTypes {
  OAS3 = 'oas3',
  SHARED = 'oas3_shared',
  TAGS = 'oas3_tags',
  SERVERS = 'oas3_servers',
  INFO = 'oas3_info',
  OPERATION = 'operation',
  COMPONENTS = 'oas3_components',
  PATHS = 'oas3_paths',
}

export enum Oas3ChildNodeTypes {
  SECURITY_SCHEME = 'oas3_security_scheme',
  PATH = 'oas3_path',
  TAG = 'oas3_tag',
  SERVER = 'oas3_server',
}

export enum OasNodeChildType {
  OBJECT = 1,
  ARRAY,
}

export enum Oas2SharedNodeTypes {
  PARAMETERS = 'oas2_parameters',
  PARAMETER = 'oas2_parameter',
  RESPONSES = 'oas2_responses',
  RESPONSE = 'oas2_response',
  JSON_SCHEMA = 'json_schema',
  SCHEMAS = 'oas2_schemas',
  OPERATION = 'oas2_operation',
}

export enum Oas3SharedNodeTypes {
  SCHEMAS = 'oas3_schemas',
  RESPONSES = 'oas3_responses',
  RESPONSE = 'oas3_response',
  PARAMETERS = 'oas3_parameters',
  EXAMPLES = 'oas3_examples',
  REQUEST_BODIES = 'oas3_request_bodies',
  HEADERS = 'oas3_headers',
  SECURITY_SCHEMES = 'oas3_security_schemes',
  LINKS = 'oas3_links',
  CALLBACKS = 'oas3_callbacks',
  OPERATION = 'oas3_operation',
}
