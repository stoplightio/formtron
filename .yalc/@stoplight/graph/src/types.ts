import { IParserResult } from '@stoplight/types';

export type IGraphLoader<Node extends INodeInstance = INodeInstance, NodeContent = any> = (
  node: Node
) => Promise<NodeContent>;

export interface IGraphHookApi<Input, Output, Parent> {
  createNode: (node: Input, opts?: { parent?: Parent }) => Promise<Output>;
  createEdge: (source: INodeInstance, destination: INodeInstance) => void;
  getNode: (id: string) => INodeInstance | undefined;
  getParentNode: (node: INodeInstance) => INodeInstance | undefined;
}

export interface IGraphHook<
  Input extends INodeOptions = INodeOptions,
  Output extends INodeInstance = INodeInstance,
  Parent extends INodeInstance = INodeInstance
> {
  name: string;

  // by default node type is set to name, but can override here if necessary
  // TODO(SL-248): do we even need this? It looks that this is only used to ensure that the produced node
  // is of the same type that what hook declares. Can't we solve that differently?
  nodeType?: string;

  // allow plugins to opt into dangerous recursive handling
  // we can put additional safeguards around this later
  recursive?: boolean;

  // scopes which nodes this plugin is interested in
  selector?: ((node: INodeInstance) => boolean) | string;

  // events
  onWillCreateNode?: (node: Partial<Output>) => void;

  // TODO(SL-248): why does that have to return anything? What is `any`?
  onDidCreateNode?: (parent: Parent, opts: IGraphHookApi<Input, Output, Parent>) => Promise<any>;
}

export interface IGraphPlugin {
  hooks?: Array<IGraphHook<any, any, any>>;
  loaders?: {
    [nodeType: string]: IGraphLoader<any, any>;
  };
}

/**
 * Nodes
 */

export enum NodeTypes {
  PARSED = 'parsed',
}

// When creating a node.
export interface INodeOptions<Content = unknown> extends Object {
  id: string;
  type: string;
  contentType?: string;
  // TODO(SL-248): I think this should be private in context of INode because INode has loadContent function
  content?: Content;
}

// When dealing with a node instance.
export interface INodeInstance<Content = unknown> extends INodeOptions<Content> {
  loadContent: () => Promise<Content>;
}

export interface IParsedNode<T> extends INodeInstance<IParserResult<T>> {
  type: NodeTypes.PARSED;
}
