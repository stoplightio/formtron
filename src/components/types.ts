export enum Variant {
  normal = '',
  invalid = 'invalid',
}
