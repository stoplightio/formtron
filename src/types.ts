import * as React from 'react';
import { Dictionary } from 'ts-essentials';
import { themeTypes } from './theme';

export interface IFormtronCommon {
  id?: string;
  value: any;
  onChange: (value: any) => void;

  schema: any;
  fieldComponents: Dictionary<React.FunctionComponent<IFormtronControl>>;
}

export interface IFormtronControl extends IFormtronCommon {
  path: string[];
}

export interface IFormtron extends IFormtronCommon {
  selection: string;
  onInternalChange?: Function;
  themeName?: themeTypes;
}

export interface IAddOperation {
  op: 'add';
  path: string;
  value: any;
}

export interface IMoveOperation {
  op: 'move';
  from: string;
  path: string;
}

export interface ISelectOperation {
  op: 'select';
  from: string;
  path: string;
}

export type IOperation = IAddOperation | IMoveOperation | ISelectOperation;
