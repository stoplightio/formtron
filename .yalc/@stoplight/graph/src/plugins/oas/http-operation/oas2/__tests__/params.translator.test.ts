import { HttpParamStyles } from '@stoplight/types';
import { QueryParameter } from 'swagger-schema-official';
import { translateToContent } from '../helpers/content.translator';
import {
  translateToBodyParameter,
  translateToFormDataParameter,
  translateToHeaderParam,
  translateToHeaderParams,
  translateToPathParameter,
  translateToQueryParameter,
} from '../helpers/params.translator';
jest.mock('../helpers/content.translator');

describe('params.translator', () => {
  const fakeTranslatedContent = 'some fake content';
  const consumes = ['*'];

  beforeEach(() => {
    (translateToContent as jest.Mock).mockReturnValue(fakeTranslatedContent);
  });

  describe('translateToHeaderParam', () => {
    test('should translate header param', () => {
      expect(
        translateToHeaderParam(
          {
            required: true,
            name: 'name',
            type: 'type',
            in: 'body',
            minimum: 12,
            description: 'desc',
          },
          consumes
        )
      ).toMatchSnapshot();
    });
  });

  describe('translateToHeaderParams', () => {
    test('should translate empty dictionary to empty array', () => {
      expect(translateToHeaderParams({}, consumes)).toMatchSnapshot();
    });

    test('should translate to simple header param', () => {
      expect(
        translateToHeaderParams(
          {
            'header-name': {
              description: 'a description',
              type: 'a type',
            },
          },
          consumes
        )
      ).toMatchSnapshot();
    });

    test('should translate to multiple header params', () => {
      expect(
        translateToHeaderParams(
          {
            'header-name': {
              description: 'a description',
              type: 'a type',
            },
            'plain-tex': {
              description: 'another description',
              type: 'another type',
            },
          },
          consumes
        )
      ).toMatchSnapshot();
    });
  });

  describe('translateToBodyParameter', () => {
    test('should translate to body parameter', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              format: 'e-mail',
            },
          },
          consumes
        )
      ).toMatchSnapshot();
    });
  });

  describe('translateToFormDataParameter', () => {
    test('given request body with empty encodings should create encoding', () => {
      expect(
        translateToFormDataParameter(
          {
            name: 'name',
            type: 'formData',
            description: 'desc',
            required: true,
            in: 'body',
          },
          {
            contents: [
              {
                mediaType: 'application/json',
              },
            ],
          },
          consumes
        )
      ).toMatchSnapshot();
    });

    test('given request body with existing encoding should append', () => {
      expect(
        translateToFormDataParameter(
          {
            name: 'name',
            type: 'formData',
            description: 'desc',
            required: true,
            in: 'body',
          },
          {
            contents: [
              {
                mediaType: 'application/json',
                encodings: [{ property: 'prop', style: HttpParamStyles.SpaceDelimited }],
              },
            ],
          },
          consumes
        )
      ).toMatchSnapshot();
    });

    test('given no request body should create one', () => {
      expect(
        translateToFormDataParameter(
          {
            name: 'name',
            type: 'formData',
            description: 'desc',
            required: true,
            in: 'body',
          },
          null,
          consumes
        )
      ).toMatchSnapshot();
    });
  });

  describe('translateToQueryParameter', () => {
    const parameter: QueryParameter = {
      required: true,
      description: 'descr',
      name: 'name',
      in: 'query',
      type: 'array',
      allowEmptyValue: true,
    };

    test.each(['pipes', 'ssv', 'csv', 'multi', 'something-else'])('translate style: %s', style => {
      expect(
        translateToQueryParameter({ ...parameter, collectionFormat: style } as QueryParameter, consumes)
      ).toMatchSnapshot();
    });
  });

  describe('translateToPathParameter', () => {
    test('should translate', () => {
      expect(
        translateToPathParameter(
          {
            required: true,
            description: 'descr',
            name: 'name',
            in: 'path',
            type: 'string',
          },
          consumes
        )
      ).toMatchSnapshot();
    });
  });
});
