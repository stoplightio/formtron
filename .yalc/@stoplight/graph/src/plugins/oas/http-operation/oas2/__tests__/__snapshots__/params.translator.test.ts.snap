// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`params.translator translateToBodyParameter should translate to body parameter 1`] = `
Object {
  "contents": Array [
    Object {
      "mediaType": "*",
      "schema": Object {
        "format": "e-mail",
      },
    },
  ],
  "description": "descr",
  "required": true,
}
`;

exports[`params.translator translateToFormDataParameter given no request body should create one 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "mediaType": "*",
      "schema": Object {
        "description": "desc",
        "in": "body",
        "name": "name",
        "required": true,
        "type": "formData",
      },
    },
  ],
  "description": "desc",
  "required": true,
}
`;

exports[`params.translator translateToFormDataParameter given request body with empty encodings should create encoding 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "mediaType": "application/json",
    },
  ],
}
`;

exports[`params.translator translateToFormDataParameter given request body with existing encoding should append 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "prop",
          "style": "spaceDelimited",
        },
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "mediaType": "application/json",
    },
  ],
}
`;

exports[`params.translator translateToHeaderParam should translate header param 1`] = `
Object {
  "content": "some fake content",
  "description": "desc",
  "name": "name",
  "required": true,
  "style": "simple",
}
`;

exports[`params.translator translateToHeaderParams should translate empty dictionary to empty array 1`] = `Array []`;

exports[`params.translator translateToHeaderParams should translate to multiple header params 1`] = `
Array [
  Object {
    "content": "some fake content",
    "description": "a description",
    "name": "header-name",
    "style": "simple",
  },
  Object {
    "content": "some fake content",
    "description": "another description",
    "name": "plain-tex",
    "style": "simple",
  },
]
`;

exports[`params.translator translateToHeaderParams should translate to simple header param 1`] = `
Array [
  Object {
    "content": "some fake content",
    "description": "a description",
    "name": "header-name",
    "style": "simple",
  },
]
`;

exports[`params.translator translateToPathParameter should translate 1`] = `
Object {
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "simple",
}
`;

exports[`params.translator translateToQueryParameter translate style: csv 1`] = `
Object {
  "allowEmptyValue": true,
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: multi 1`] = `
Object {
  "allowEmptyValue": true,
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: pipes 1`] = `
Object {
  "allowEmptyValue": true,
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "pipeDelimited",
}
`;

exports[`params.translator translateToQueryParameter translate style: something-else 1`] = `
Object {
  "allowEmptyValue": true,
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: ssv 1`] = `
Object {
  "allowEmptyValue": true,
  "content": "some fake content",
  "description": "descr",
  "name": "name",
  "required": true,
  "style": "spaceDelimited",
}
`;
